README
NAME
    Dancer::Plugin::REST - A plugin for writing RESTful apps with Dancer

VERSION
    version 0.08

DESCRIPTION
    This plugin helps you write a RESTful webservice with Dancer.

SYNOPSYS
        package MyWebService;

        use Dancer;
        use Dancer::Plugin::REST;

        prepare_serializer_for_format;

        get '/user/:id.:format' => sub {
            User->find(params->{id});
        };

        # curl http://mywebservice/user/42.json
        { "id": 42, "name": "John Foo", email: "john.foo@example.com"}

        # curl http://mywebservice/user/42.yml
        --
        id: 42
        name: "John Foo"
        email: "john.foo@example.com"

KEYWORDS
  prepare_serializer_for_format
    When this pragma is used, a before filter is set by the plugin to
    automatically change the serializer when a format is detected in the
    URI.

    That means that each route you define with a :format token will trigger
    a serializer definition, if the format is known.

    This lets you define all the REST actions you like as regular Dancer
    route handlers, without explicitly handling the outgoing data format.

  resource
    This keyword lets you declare a resource your application will handle.

        resource user =>
            get    => sub { # return user where id = params->{id}   },
            create => sub { # create a new user with params->{user} },
            delete => sub { # delete user where id = params->{id}   },
            update => sub { # update user with params->{user}       };

        # this defines the following routes:
        # GET /user/:id
        # GET /user/:id.:format
        # POST /user
        # POST /user.:format
        # DELETE /user/:id
        # DELETE /user/:id.:format
        # PUT /user/:id
        # PUT /user/:id.:format

  helpers
    Some helpers are available. This helper will set an appropriate HTTP
    status for you.

   status_ok
        status_ok({users => {...}});

    Set the HTTP status to 200

   status_created
        status_created({users => {...}});

    Set the HTTP status to 201

   status_accepted
        status_accepted({users => {...}});

    Set the HTTP status to 202

   status_bad_request
        status_bad_request("user foo can't be found");

    Set the HTTP status to 400. This function as for argument a scalar that
    will be used under the key error.

   status_not_found
        status_not_found("users doesn't exists");

    Set the HTTP status to 404. This function as for argument a scalar that
    will be used under the key error.

LICENCE
    This module is released under the same terms as Perl itself.

AUTHORS
    This module has been written by Alexis Sukrieh "<sukria@sukria.net>" and
    Franck Cuny.

SEE ALSO
    Dancer <http://en.wikipedia.org/wiki/Representational_State_Transfer>

AUTHORS
    *   Alexis Sukrieh <sukria@sukria.net>

    *   Franck Cuny <franckc@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Alexis Sukrieh.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
$NEXT

    [BUG FIXES]
    * 'prepare_serializer_for_format' was clobbering the
        default serializer. (RT#93478, reported by Pierre Vigier)

0.07    2011-11-01

    * Minor updates to fix warnings with Dancer 1.3080
    * Fix POD
    (Alexis Sukrieh)

0.06	2011-05-28

	[ TEST SUITE ]
	* Skip test t/03_resource.t if Dancer is not >= 1.3059_01
	  A bug in Dancer::Test in previous versions prevent the test from passing.

0.05    2011-03-04

    [ BUG FIXES ]
    * Fix the test suite with dancer_response
    * Fix dependencies: Dancer 1.3014

Version 0.04
	[ Alexis Sukrieh ]
	* Add all HTTP response status

Version 0.03
    [ Franck Cuny ]
    * New keywords for ReST response status 
	  (status_ok, status_created, ...)

Version 0.02
	* 100% coverage
	* fix a bug in 'resource' (order of route hander definition)

Version 0.0001_01
	* Support for prepare_serializer_for_format keyword
	  autoamtically set a serializer according to params->{format}
	* Support for resource definition
	  When a resource is defined a set of routes are defined in a RESTful manner.


lib/Dancer/Plugin/REST.pm
package Dancer::Plugin::REST;
use strict;
use warnings;

use Carp 'croak';
use Dancer ':syntax';
use Dancer::Plugin;

our $AUTHORITY = 'SUKRIA';
our $VERSION   = '0.07';

my $content_types = {
    json => 'application/json',
    yml  => 'text/x-yaml',
    xml  => 'application/xml',
};

our $default_serializer;

register prepare_serializer_for_format => sub {
    my $conf        = plugin_setting;
    my $serializers = (
        ($conf && exists $conf->{serializers})
        ? $conf->{serializers}
        : { 'json' => 'JSON',
            'yml'  => 'YAML',
            'xml'  => 'XML',
            'dump' => 'Dumper',
        }
    );

    hook 'before' => sub {

        my $format = params->{'format'}
            or return;

        # remember what was there before
        $default_serializer = setting 'serializer';

        my $serializer = $serializers->{$format}
            or return halt(
                Dancer::Error->new(
                    code    => 404,
                    title   => "unsupported format requested",
                    message => "unsupported format requested: " . $format
                )->render
            );

        set serializer => $serializer;

        content_type $content_types->{$format} || setting('content_type');
    };

    hook after => sub {
        # put it back the way it was
        set serializer => $default_serializer if $default_serializer;
    }
};

register resource => sub {
    my ($resource, %triggers) = @_;

    croak "resource should be given with triggers"
      unless defined $resource
          and defined $triggers{get}
          and defined $triggers{update}
          and defined $triggers{delete}
          and defined $triggers{create};

    get "/${resource}/:id.:format" => $triggers{get};
    get "/${resource}/:id"         => $triggers{get};

    put "/${resource}/:id.:format" => $triggers{update};
    put "/${resource}/:id"         => $triggers{update};

    post "/${resource}.:format" => $triggers{create};
    post "/${resource}"         => $triggers{create};

    del "/${resource}/:id.:format" => $triggers{delete};
    del "/${resource}/:id"         => $triggers{delete};
};

register send_entity => sub {
    my ($entity, $http_code) = @_;

    $http_code ||= 200;

    status($http_code);
    $entity;
};

my %http_codes = (

    # 1xx
    100 => 'Continue',
    101 => 'Switching Protocols',
    102 => 'Processing',

    # 2xx
    200 => 'OK',
    201 => 'Created',
    202 => 'Accepted',
    203 => 'Non-Authoritative Information',
    204 => 'No Content',
    205 => 'Reset Content',
    206 => 'Partial Content',
    207 => 'Multi-Status',
    210 => 'Content Different',

    # 3xx
    300 => 'Multiple Choices',
    301 => 'Moved Permanently',
    302 => 'Found',
    303 => 'See Other',
    304 => 'Not Modified',
    305 => 'Use Proxy',
    307 => 'Temporary Redirect',
    310 => 'Too many Redirect',

    # 4xx
    400 => 'Bad Request',
    401 => 'Unauthorized',
    402 => 'Payment Required',
    403 => 'Forbidden',
    404 => 'Not Found',
    405 => 'Method Not Allowed',
    406 => 'Not Acceptable',
    407 => 'Proxy Authentication Required',
    408 => 'Request Time-out',
    409 => 'Conflict',
    410 => 'Gone',
    411 => 'Length Required',
    412 => 'Precondition Failed',
    413 => 'Request Entity Too Large',
    414 => 'Request-URI Too Long',
    415 => 'Unsupported Media Type',
    416 => 'Requested range unsatisfiable',
    417 => 'Expectation failed',
    418 => 'Teapot',
    422 => 'Unprocessable entity',
    423 => 'Locked',
    424 => 'Method failure',
    425 => 'Unordered Collection',
    426 => 'Upgrade Required',
    449 => 'Retry With',
    450 => 'Parental Controls',

    # 5xx
    500 => 'Internal Server Error',
    501 => 'Not Implemented',
    502 => 'Bad Gateway',
    503 => 'Service Unavailable',
    504 => 'Gateway Time-out',
    505 => 'HTTP Version not supported',
    507 => 'Insufficient storage',
    509 => 'Bandwidth Limit Exceeded',
);

for my $code (keys %http_codes) {
    my $helper_name = lc($http_codes{$code});
    $helper_name =~ s/[^\w]+/_/gms;
    $helper_name = "status_${helper_name}";

    register $helper_name => sub {
        if ($code >= 400) {
            send_entity({error => $_[0]}, $code);
        }
        else {
            send_entity($_[0], $code);
        }
    };
}

register_plugin;
1;
__END__

=pod

=head1 NAME

Dancer::Plugin::REST - A plugin for writing RESTful apps with Dancer

=head1 SYNOPSYS

    package MyWebService;

    use Dancer;
    use Dancer::Plugin::REST;

    prepare_serializer_for_format;

    get '/user/:id.:format' => sub {
        User->find(params->{id});
    };

    # curl http://mywebservice/user/42.json
    { "id": 42, "name": "John Foo", email: "john.foo@example.com"}

    # curl http://mywebservice/user/42.yml
    --
    id: 42
    name: "John Foo"
    email: "john.foo@example.com"

=head1 DESCRIPTION

This plugin helps you write a RESTful webservice with Dancer.

=head1 KEYWORDS

=head2 prepare_serializer_for_format

When this pragma is used, a before filter is set by the plugin to automatically
change the serializer when a format is detected in the URI.

That means that each route you define with a B<:format> token will trigger a
serializer definition, if the format is known.

This lets you define all the REST actions you like as regular Dancer route
handlers, without explicitly handling the outgoing data format.

=head2 resource

This keyword lets you declare a resource your application will handle.

    resource user =>
        get    => sub { # return user where id = params->{id}   },
        create => sub { # create a new user with params->{user} },
        delete => sub { # delete user where id = params->{id}   },
        update => sub { # update user with params->{user}       };

    # this defines the following routes:
    # GET /user/:id
    # GET /user/:id.:format
    # POST /user
    # POST /user.:format
    # DELETE /user/:id
    # DELETE /user/:id.:format
    # PUT /user/:id
    # PUT /user/:id.:format

=head2 helpers

Some helpers are available. This helper will set an appropriate HTTP status for you.

=head3 status_ok

    status_ok({users => {...}});

Set the HTTP status to 200

=head3 status_created

    status_created({users => {...}});

Set the HTTP status to 201

=head3 status_accepted

    status_accepted({users => {...}});

Set the HTTP status to 202

=head3 status_bad_request

    status_bad_request("user foo can't be found");

Set the HTTP status to 400. This function as for argument a scalar that will be used under the key B<error>.

=head3 status_not_found

    status_not_found("users doesn't exists");

Set the HTTP status to 404. This function as for argument a scalar that will be used under the key B<error>.

=head1 LICENCE

This module is released under the same terms as Perl itself.

=head1 AUTHORS

This module has been written by Alexis Sukrieh C<< <sukria@sukria.net> >> and Franck
Cuny.

=head1 SEE ALSO

L<Dancer> L<http://en.wikipedia.org/wiki/Representational_State_Transfer>

=cut
